### PR TITLE
feat: add colorpicker setting element

### PR DESCRIPTION
**Pull request by TaroEld**
Adds a color picker element. Also improves / standardises CSS of the setting elements and fixes some persistent data issues.

### CODE REVIEW

File: msu/utilities/exceptions.nut [hunk 1]
```diff
@@ -8,7 +8,7 @@ local function getInQuotes( _string )
 }
 
 ::MSU.Exception <- {
-	KeyNotFound = function( _string ) { return format("Key%s not found", getinQuotes(_string)); },
+	KeyNotFound = function( _string ) { return format("Key%s not found", getInQuotes(_string)); },
```
Comment (Enduriel) [inline msu/utilities/exceptions.nut:11]: LOL

File: scripts/mods/msu/settings_screen.nut [hunk 2]
```diff
@@ -35,6 +35,7 @@ this.settings_screen <- ::inherit("scripts/mods/msu/ui_screen", {
 	function connect()
 	{
 		this.m.JSHandle = ::UI.connect("ModSettingsScreen", this);
+		this.m.JSHandle.asyncCall("setupJSSettings", ::MSU.System.ModSettings.getUIData(null));
```
Comment (Enduriel) [inline scripts/mods/msu/settings_screen.nut:38]: remove since this already exists

File: ui/mods/msu/mod_settings/color_picker_setting.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
+var ColorPickerSetting = function (_mod, _page, _setting, _parentDiv)
+{
+
```
Comment (Enduriel) [inline ui/mods/msu/mod_settings/color_picker_setting.js:3]: remove newline

File: ui/mods/msu/mod_settings/color_picker_setting.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		self.cloneSettingsToFrom(self.values, self.popup.values);
+		self.updateColor();
+		_dialog.destroyPopupDialog();
+	})
+	this.popup.addPopupDialogCancelButton(function (_dialog){
+		_dialog.destroyPopupDialog();
+	})
+	this.popup.addPopupDialogContent(this.createPopupContent())
+	this.updateColorInPopup();
+
```
Comment (Enduriel) [inline ui/mods/msu/mod_settings/color_picker_setting.js:58]: remove newline

File: ui/mods/msu/mod_settings/color_picker_setting.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+ColorPickerSetting.prototype.getRGBA = function( _values )
+{
+	var result = "rgba(";
+	result += _values.Red + ", ";
+	result += _values.Green + ", ";
+	result += _values.Blue + ", ";
+	result += _values.Alpha + ");";
+	return result
+}
```
Comment (Enduriel) [inline ui/mods/msu/mod_settings/color_picker_setting.js:129]: Not sure why these functions are static

File: ui/mods/msu/mod_settings/settings_screen.js [hunk 6]
```diff
@@ -1,3 +1,4 @@
+var ModSettings = {};
```
Comment (Enduriel) [inline ui/mods/msu/mod_settings/settings_screen.js:1]: remove all the references to these new js mod-settings

File: ui/mods/msu/mod_settings/settings_screen.js [hunk 7]
```diff
@@ -264,9 +274,22 @@ ModSettingsScreen.prototype.getChanges = function () // Could still be significa
 			});
 		});
 	});
+	ModSettings = changes;
```
Comment (Enduriel) [inline ui/mods/msu/mod_settings/settings_screen.js:277]: remove

**Review by Enduriel — CHANGES_REQUESTED** [review]

File: ui/mods/msu/mod_settings/color_picker_setting.js (hunk 5 above)
Comment (TaroEld) [reply ui/mods/msu/mod_settings/color_picker_setting.js:130]: There's only ever going to be these entries. It shows you exactly what's happening. No need to iterate over the object here-

File: msu/systems/mod_settings/abstract_setting.nut [hunk 8]
```diff
@@ -30,7 +30,12 @@
 
 	function printForParser( _tag = "ModSetting" )
 	{
-		::MSU.System.PersistentData.writeToLog(_tag, this.getModID(), [this.getID(), this.getValue().tostring()]);
+		local payload = this.getValue();
+		if (typeof payload == "string")
+		{
+			payload = "\"" + payload + "\"";
+		}
+		::MSU.System.PersistentData.writeToLog(_tag, this.getModID(), [this.getID(), payload]);
```
[the remark this answers is not in the document]
Comment (TaroEld) [reply msu/systems/mod_settings/abstract_setting.nut:38]: It's a way to get quotes around a string

File: msu/systems/mod_settings/color_picker_setting.nut [hunk 9]
```diff
@@ -0,0 +1,71 @@
+::MSU.Class.ColorPickerSetting <- class extends ::MSU.Class.AbstractSetting
+{
+	ValuesAsTable = null;
+	ValuesAsRGBA = null;
```
[the remark this answers is not in the document]
Comment (TaroEld) [reply msu/systems/mod_settings/color_picker_setting.nut:4]: fair enough

**Comment by Enduriel** [comment]
Was split up into individual PRs for the individual features/fixes/refactors

**Pull request closed without merging**